### ISSUE TITLE
(M2C) (PC) check_existence method in pohandler (blocked by #353, #384)

### ISSUE THREAD

**Issue by ndushay** [sul-dlss/preservation_catalog#284]
check_existence method will:

- [ ] call confirm_version
- [ ] if object not found error, calls create_after_validation
- [ ] if Moab version > catalog, calls update_version_after_validation
- [ ] else error

(timestamp info also in #286)

PC timestamp changes:
if check_existence clean:
- [ ] last_checked_on_storage AND 
- [ ] last_audited AND 
- [ ] updated_at (but ActiveRecord does this, right?)

PO timestamp(s) changes only when check_existence:
- calls create_after_validation (already done in plain create method, timestamp info in ticket #288)
- calls update_version_after_validation (timestamp info already in ticket #287)

- [ ] lib/audit code will call check_existence method instead of confirm_version

**Comment by ndushay** [comment]
This spec needs to be confirmed against the appropriate doc (https://docs.google.com/document/d/1QdsQyR40l6_ykcibJ4OxxFCs1-KiGmdTzzpw0J2Telg ??  ):

calls confirm_version
- if versions equal - yay.  update timestamps or whatever as appropriate.
- if version mismatch (moab > catalog), calls update_version_after_validation
- if version mismatch (catalog > moab), alert, etc.
- if moab not in catalog:  create_after_validaiton

timestamp ticket #286

**Comment by ndushay** [comment]
possibly blocked by #353

**Comment by SaravShah** [comment]
M2c in audit will call this method

**Comment by SaravShah** [comment]
Move meat of M2C to a method in the POH

**Comment by jmartin-sul** [comment]
paired w/ @tallenaz on this on friday, and we mostly walked through the existing code to try and figure out what needed to be written.  here's what we came up with:
* seems like once WIP PR #379 is in (refactoring confirm_version to look more like update_version), with the validation for version mismatch, then we'd have most of the behavior that's outlined in the audit requirements doc by calling current M2C `check_existence` method.
  * except that we'd need to call `create_after_validation` instead of `create` when the object doesn't exist
* as such, feels like the first thing that's needed is to have `check_existence` (or its POH version) call `create_after_validation` (rebased on the WIP PR if M2C `check_existence` were to be used).
* also seems that we need to set validation timestamps (last audited, last checked on storage).  since it seems like `confirm_version` doesn't do this (either on master or the version in the PR).
  * could have a `confirm_version_after_validation` method, analagous to `update_version_after_validation`
    * in which case, part of the WIP refactor would be unneeded: instead of calling `moab_validation_errors` in `confirm_online_version`, we could pass `confirm_online_version` a validated flag (and a status?), similar to what `update_online_version` does.
    * less clear where we'd do validation if we kept the same structure for `confirm_version` and its helpers as we're currently using in master.
* then, there's still the issue of where this `check_existence` method lives: M2C or POH.  this ticket and F2F discussions implied to me that @ndushay votes POH since it could be a rest call.  but couldn't a rest call just as easily use M2C?  or, is naomi's point that there's a version of check_existence in M2C as well as in POH?
  * FWIW, none of the current check existence/confirm version machinery appears to be used by anything other than the tests, at this point (e.g. no rake task calling that).
* regardless of where it lives, also need a rake task to invoke the check, right?  prob have the rake task take an optional storage root, and call check_existence for a single root method, for easier parallelization?

**Comment by jmartin-sul** [comment]
we were going off this ticket, the linked tickets in the description, and this audit requirements doc:  https://docs.google.com/document/d/1QdsQyR40l6_ykcibJ4OxxFCs1-KiGmdTzzpw0J2Telg/edit#heading=h.qg87o9xyb5gn

**Comment by jmartin-sul** [comment]
clarification from @ndushay on where the code lives:  the overall `check_existence` code for a storage root (or for all the storage roots) continues to live in M2C.  the code in its inner loop, for dealing with a single moab, should be extracted out to POH.

**Comment by ndushay** [comment]
Decisions:
- confirm_version will be "stupider", and will only do the "right' thing for matching version happy path (results, timestamps), o.w. it will error
- pohandler #check_existence will call confirm_version.  Either confirm_version does happy path and we're done (?) or we suss out if we have an error we can work with and continue per first description in this ticket.
    - some of this logic may currently live in lib/audit/m2c, but that method will become stupider and the pohandler method will become the place for logic.
